src/s390_crypto.c

```diff
@@ -318,14 +325,114 @@ static void set_switches(int msa)
 			*s390_kdsa_functions[n].enabled = 1;
 }
 
+unsigned int is_device_online(const char *dev)
+{
+	unsigned int ret = 0;
+	FILE *file;
+	char c;
+
+	if ((file = fopen(dev, "r")) == NULL)
+		return 0;
+
+	/* Check if device online: 0 = offline, 1 = online */
+	if ((c = fgetc(file)) == '1')
+		ret = 1;
+
+	fclose(file);
+
+	return ret;
+}
+
+unsigned int get_device_type(const char *dev, char *devtype)
+{
+	unsigned int ret = 0;
+	char *type = NULL;
+	size_t size = 0;
+	FILE *file;
+
+	file = fopen(dev, "r");
+	if (file == NULL)
+		return 0;
+
+	/* Read device type: e.g. 'CEX5C' */
+	if (getline(&type, &size, file) == -1)
+		goto end;
+
+	/* Did we get something like 'CEX5C\n'? */
+	if (strlen(type) != 6 || strstr(type, "CEX") != &type[0])
+		goto end;
+
+	memcpy(devtype, type, 5);
+	ret = 1;
+
+end:
+	free(type);
+	fclose(file);
+
+	return ret;
+}
+
+#define AP_PATH  "/sys/devices/ap"
+#define MAX_DEV_LEN 280
+
+unsigned int search_for_cards()
+{
+	DIR *sysDir;
+	unsigned int ret = 0;
+	char dev[MAX_DEV_LEN] = AP_PATH;
+	struct dirent *direntp;
+	char type[6];
+
+	if ((sysDir = opendir(dev)) == NULL)
+		return 0;
+
+	while ((direntp = readdir(sysDir)) != NULL) {
+
+		/* Skip entries that are not like "card01", "card02", etc. */
+		if (strstr(direntp->d_name, "card") != &direntp->d_name[0])
+			continue;
+
+		/* Check if device online */
+		snprintf(dev, MAX_DEV_LEN, "%s/%s/online", AP_PATH, direntp->d_name);
+		if (!is_device_online(dev))
+			continue;
+
+		/* Get device type (string like "CEXnT") */
+		snprintf(dev, MAX_DEV_LEN, "%s/%s/type", AP_PATH, direntp->d_name);
+		memset(type, 0, sizeof(type));
```

Comment:
the memset here is not needed as you check the returncode from get_device_type()

Reply:
Yes, that's right.